### ISSUE TITLE
Data.List.upTo/applyUpTo can be made faster

### ISSUE THREAD

**Issue by Taneb** [agda/agda-stdlib#2437]
In a quick experiment, the following definition runs circles around the current implementation of `upTo`

```agda
goUpTo : ℕ → ℕ → List ℕ
goUpTo _ 0 = []
goUpTo n (suc i) = n ∷ goUpTo (suc n) i

upTo′ : ℕ → List ℕ
upTo′ = goUpTo 0
```

I suspect this is because `applyUpTo` builds huge closures.

**Comment by jamesmckinna** [comment]
Can it be made faster still via a tail-call optimisation?

**Comment by jamesmckinna** [comment]
(Deleted my original comment, but now reinstating it)
Can the implementation be further improved/speeded-up by considering a tail-call optimisation?

**Comment by Taneb** [comment]
>  Can the implementation be further improved/speeded-up by considering a tail-call optimisation?

I don't think so. The recursion here is already guarded behind a constructor. I don't know what a tail recursive implementation would look like.

**Comment by linyue0103** [comment]
#679 _safe

module base where

open import Funnction public using(id:$$_-_**）

Eq public using（—=——；refl)

**Comment by jamesmckinna** [comment]
@linyue0103 not sure I understand your comment, nor how it applies to this issue?

**Comment by MatthewDaggitt** [comment]
Yes, I never had performance in mind when I add any of my definitions. Apologies! There's probably plenty more of these lurking in the library...

**Comment by jamesmckinna** [comment]
So...: instead have
```agda
upTo : ℕ → List ℕ
upTo = go 0
  module UpTo where
    go : ℕ → ℕ → List ℕ
    go _ zero    = []
    go n (suc i) = n ∷ go (suc n) i

downFrom : ℕ → List ℕ
downFrom zero    = []
downFrom (suc n) = n ∷ downFrom n
```

to avoid building the closures in `applyUpTo`/`applyDownFrom`

`allFin` seems trickier, because the closures, even if only specialised to powers of `suc`, do seem necessary?

Not clear what the knock-on effects on `Data.List.Properties` might be, but ...
